File: pi42/exchange.py
"""
Exchange API endpoints for Pi42.
"""

from typing import Dict, Any, Optional


class ExchangeAPI:
    """
    Exchange API client for managing Pi42 exchange settings.
    """

    def __init__(self, client):
        """
        Initialize the Exchange API client.

        Args:
            client: The main Pi42Client instance
        """
        self._client = client

    def get_exchange_info(self, market: Optional[str] = None) -> Dict[str, Any]:
        """
        Retrieve exchange information.

        Args:
            market: Market to return exchangeInfo for (e.g., 'INR' or 'USDT')

        Returns:
            Dict: Exchange information
        """
        endpoint = "/v1/exchange/exchangeInfo"
        
        params = {}
        if market:
            params["market"] = market
            
        return self._client.get_request(endpoint, params)

    def update_preference(self, leverage: int, margin_mode: str, 
                         contract_name: str) -> Dict[str, Any]:
        """
        Update the leverage and margin-mode for a specified contract.

        Args:
            leverage: The leverage level to set for the contract (must be an integer)
            margin_mode: Margin mode to set for the contract ('CROSS' or 'ISOLATED')
            contract_name: The trading pair or contract name (e.g., 'BTCINR')

        Returns:
            Dict: Preference update response
        """
        endpoint = "/v1/exchange/update/preference"
        
        params = {
            "leverage": leverage,
            "marginMode": margin_mode,
            "contractName": contract_name
        }
        
        return self._client.post_request(endpoint, params)

    def update_leverage(self, leverage: int, contract_name: str) -> Dict[str, Any]:
        """
        Update the leverage for a specified contract.

        Args:
            leverage: The leverage level to set for the contract (must be an integer)
            contract_name: The trading pair or contract name (e.g., 'BTCINR')

        Returns:
            Dict: Leverage update response
        """
        endpoint = "/v1/exchange/update/leverage"
        
        params = {
            "leverage": leverage,
            "contractName": contract_name
        }
        
        return self._client.post_request(endpoint, params)

File: exchange.go
package pi42

import (
	"encoding/json"
	"fmt"
)

// ExchangeAPI provides access to exchange settings endpoints
type ExchangeAPI struct {
	client *Client
}

// NewExchangeAPI creates a new Exchange API instance
func NewExchangeAPI(client *Client) *ExchangeAPI {
	return &ExchangeAPI{client: client}
}

// ExchangeInfo retrieves exchange information with structured response
func (api *ExchangeAPI) ExchangeInfo(market string) (*ExchangeInfoResponse, error) {
	endpoint := "/v1/exchange/exchangeInfo"

	params := make(map[string]string)
	if market != "" {
		params["market"] = market
	}

	data, err := api.client.Get(endpoint, params, true)
	if err != nil {
		return nil, err
	}

	var result ExchangeInfoResponse
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return &result, nil
}

// UpdatePreference updates the leverage and margin-mode for a specified contract
func (api *ExchangeAPI) UpdatePreference(leverage int, marginMode, contractName string) (*PreferenceUpdateResponse, error) {
	endpoint := "/v1/exchange/update/preference"

	params := map[string]interface{}{
		"leverage":     leverage,
		"marginMode":   marginMode,
		"contractName": contractName,
	}

	data, err := api.client.Post(endpoint, params, false)
	if err != nil {
		return nil, err
	}

	var result PreferenceUpdateResponse
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return &result, nil
}

// UpdateLeverage updates the leverage for a specified contract
func (api *ExchangeAPI) UpdateLeverage(leverage int, contractName string) (*LeverageUpdateResponse, error) {
	endpoint := "/v1/exchange/update/leverage"

	params := map[string]interface{}{
		"leverage":     leverage,
		"contractName": contractName,
	}

	data, err := api.client.Post(endpoint, params, false)
	if err != nil {
		return nil, err
	}

	var result LeverageUpdateResponse
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return &result, nil
}
